File: rotate_list_61_22_03_12/main.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil || k == 0 {
		return head
	}
	var count int
	var tail *ListNode
	for p := head; p != nil; p = p.Next {
		count++
		tail = p
	}
	tail.Next = head
	k = count - (k % count )
	for i := 0; i < k; i++ {
		tail = tail.Next
	}
	head = tail.Next
	tail.Next = nil
	return head
}

File: rotate_list_61_22_03_12/src/main.rs
fn main() {
    println!("Hello, world!");
}

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution {}

impl Solution {
    pub fn rotate_right(head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
        let mut len = 0;
        let mut walk = 0;

        match head {
            Some(head1) => {
                let mut node = &head1;
                while node.next.is_some() {
                    len += 1;
                    node = node.next.as_ref().unwrap();
                }
                walk = len - (k % len);
                node.next = head;

                let mut node = head1;
                let mut prv = &Box::new(ListNode::new(-1000));
                for _ in 1..walk {
                    // prv = &node;
                    node = node.next.unwrap();
                }
                // prv.next = None;
                return None;
            }
            None => {
                return None;
            }
        }
    }
}
